updateEntity call fails when the table has system-generated columns

When using Oracle (tested with version 12c) `AbstractDatabaseEngine#updateEntity` fails when it tries to drop a system-generated column (name starts with "SYS_") from a table.

I'm not sure what causes these system columns to be added though.

The offending SQL query:

```
OracleEngine.dropColumn:519 ALTER TABLE "FDZ_APP_WORKFLOW_ELEMENT" DROP ( "SYS_STS348AI90XHRJWNMOT2K_PHJ#" )
```

And the exception stack trace:

```
com.feedzai.commons.sql.abstraction.engine.DatabaseEngineException: Error dropping column
at com.feedzai.commons.sql.abstraction.engine.impl.OracleEngine.dropColumn(OracleEngine.java:526)
at com.feedzai.commons.sql.abstraction.engine.AbstractDatabaseEngine.updateEntity(AbstractDatabaseEngine.java:517)
[...]
Caused by: java.sql.SQLException: ORA-12996: cannot drop system-generated virtual column
[...]
```

--- a/src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/OracleEngine.java
+++ b/src/main/java/com/feedzai/commons/sql/abstraction/engine/impl/OracleEngine.java
@@ -820,11 +820,21 @@ public class OracleEngine extends AbstractDatabaseEngine {
             rsColumns = s.executeQuery(String.format("SELECT COLUMN_NAME, DATA_TYPE, DATA_PRECISION FROM ALL_TAB_COLS WHERE TABLE_NAME = '%s' AND OWNER = UPPER('%s')", name, properties.getProperty(PdbProperties.USERNAME)));
 
             while (rsColumns.next()) {
-                final String dataPrecision = rsColumns.getString("DATA_PRECISION");
+                String columnName = rsColumns.getString("COLUMN_NAME");
+                /*
+                Columns starting with SYS_ are Oracle system-generated columns
+                and PDB should not interfere with these columns. Not storing
+                these in the metaMap makes them invisible to PDB.
 
-                final DbColumnType value = toPdbType(dataPrecision == null ? rsColumns.getString("DATA_TYPE") : (rsColumns.getString("DATA_TYPE") + dataPrecision));
-                if (value != DbColumnType.UNMAPPED) {
-                    metaMap.put(rsColumns.getString("COLUMN_NAME"), value);
+                See: http://docs.oracle.com/cd/B19306_01/server.102/b14200/ap_keywd.htm
+                 */
+                if (!columnName.toUpperCase().startsWith("SYS_")) {
+                    final String dataPrecision = rsColumns.getString("DATA_PRECISION");
+
+                    final DbColumnType value = toPdbType(dataPrecision == null ? rsColumns.getString("DATA_TYPE") : (rsColumns.getString("DATA_TYPE") + dataPrecision));
+                    if (value != DbColumnType.UNMAPPED) {
+                        metaMap.put(columnName, value);
+                    }
                 }
             }
 

--- a/src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/oracle/OracleEngineSchemaTest.java
+++ b/src/test/java/com/feedzai/commons/sql/abstraction/engine/impl/oracle/OracleEngineSchemaTest.java
@@ -16,16 +16,23 @@
 package com.feedzai.commons.sql.abstraction.engine.impl.oracle;
 
 
+import com.feedzai.commons.sql.abstraction.ddl.DbEntity;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseEngine;
+import com.feedzai.commons.sql.abstraction.engine.DatabaseFactory;
 import com.feedzai.commons.sql.abstraction.engine.impl.abs.AbstractEngineSchemaTest;
 import com.feedzai.commons.sql.abstraction.engine.testconfig.DatabaseConfiguration;
 import com.feedzai.commons.sql.abstraction.engine.testconfig.DatabaseTestUtil;
+import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
 
 import java.util.Collection;
 import java.util.Properties;
 
+import static com.feedzai.commons.sql.abstraction.ddl.DbColumnType.*;
+import static com.feedzai.commons.sql.abstraction.dml.dialect.SqlBuilder.*;
 import static com.feedzai.commons.sql.abstraction.engine.configuration.PdbProperties.*;
+import static org.junit.Assert.*;
 
 /**
  * @author Rafael Marmelo (rafael.marmelo@feedzai.com)
@@ -65,5 +72,32 @@ public class OracleEngineSchemaTest extends AbstractEngineSchemaTest {
     @Override
     protected String getSchema() {
         return "";
+    }
+
+    /**
+     * Checks that system generated columns with name starting with SYS_ in the
+     * Oracle engine do not appear in the table metadata.
+     *
+     * @throws Exception propagates any Exception thrown by the test
+     */
+    @Test
+    public void testSystemGeneratedColumns() throws Exception {
+        DatabaseEngine engine = DatabaseFactory.getConnection(properties);
+
+        try {
+            DbEntity entity = dbEntity()
+                    .name("TEST_SYS_COL")
+                    // Simulates a system generated column
+                    .addColumn("SYS_COL1", INT)
+                    .addColumn("COL1", INT)
+                    .pkFields("COL1")
+                    .build();
+            engine.addEntity(entity);
+
+            assertFalse("The simulated system generated column should not appear in the table metadata", engine.getMetadata("TEST_SYS_COL").containsKey("SYS_COL1"));
+            assertTrue("The regular column should appear in the table metadata", engine.getMetadata("TEST_SYS_COL").containsKey("COL1"));
+        } finally {
+            engine.close();
+        }
     }
 }

